test/matrix_test.go

```diff
@@ -0,0 +1,463 @@
+		},
+		Status: v1.TaskRunStatus{
+			TaskRunStatusFields: v1.TaskRunStatusFields{
+				Results: []v1.TaskRunResult{{
+					Name:  "GOARCHs",
+					Type:  "array",
+					Value: v1.ParamValue{Type: v1.ParamTypeArray, ArrayVal: []string{"linux/amd64", "linux/ppc64le"}},
+				}},
+			},
+			Status: duckv1.Status{Conditions: []apis.Condition{{
+				Type:    apis.ConditionSucceeded,
+				Status:  "True",
+				Reason:  "Succeeded",
+				Message: "All Steps have completed executing",
+			}}},
+		},
+	}, {
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "pr-pt2-with-result",
+		},
+		Spec: v1.TaskRunSpec{
+			ServiceAccountName: "default",
+			TaskRef:            &v1.TaskRef{Name: "task2withresults", Kind: v1.NamespacedTaskKind},
+		},
+		Status: v1.TaskRunStatus{
+			TaskRunStatusFields: v1.TaskRunStatusFields{
+				Results: []v1.TaskRunResult{{
+					Name:  "versions",
+					Type:  "array",
+					Value: v1.ParamValue{Type: v1.ParamTypeArray, ArrayVal: []string{"go1.17", "go1.18.1"}},
+				}},
+			},
+			Status: duckv1.Status{Conditions: []apis.Condition{{
+				Type:    apis.ConditionSucceeded,
+				Status:  "True",
+				Reason:  "Succeeded",
+				Message: "All Steps have completed executing",
+			}}},
+		},
+	}}
+
+	t.Logf("Waiting for PipelineRun %s in namespace %s to complete", prName, namespace)
+	if err := WaitForPipelineRunState(ctx, c, prName, timeout, PipelineRunSucceed(prName), "PipelineRunSuccess", v1Version); err != nil {
+		t.Fatalf("Error waiting for PipelineRun %s to finish: %s", prName, err)
+	}
+
+	actualTaskrunList, err := c.V1TaskRunClient.List(ctx, metav1.ListOptions{LabelSelector: fmt.Sprintf("tekton.dev/pipelineRun=%s", prName)})
+	if err != nil {
+		t.Fatalf("Error listing TaskRuns for PipelineRun %s: %s", prName, err)
+	}
+
+	actualTaskRuns := actualTaskrunList.Items
+	ignoreTypeMeta := cmpopts.IgnoreFields(metav1.TypeMeta{}, "Kind", "APIVersion")
+	ignoreObjectMeta := cmpopts.IgnoreFields(metav1.ObjectMeta{}, "Name", "Namespace", "UID", "Generation", "CreationTimestamp", "OwnerReferences", "ManagedFields", "Labels", "ResourceVersion", "Annotations")
+	ignoreTaskRunStatusFields := cmpopts.IgnoreFields(v1.TaskRunStatusFields{}, "StartTime", "CompletionTime", "PodName", "Steps", "TaskSpec", "Provenance", "Sidecars")
+	ignoreLastTransitionTime := cmpopts.IgnoreFields(apis.Condition{}, "LastTransitionTime.Inner.Time")
+	sortTaskRuns := cmpopts.SortSlices(func(x, y *v1.TaskRun) bool { return x.Name < y.Name })
+	ignoreSericeAccountName := cmpopts.IgnoreFields(v1.TaskRunSpec{}, "ServiceAccountName")
+	ignoreTimeout := cmpopts.IgnoreFields(v1.TaskRunSpec{}, "Timeout")
+
+	if d := cmp.Diff(expectedTaskRuns, actualTaskRuns, ignoreLastTransitionTime, ignoreObjectMeta, ignoreTypeMeta, ignoreTaskRunStatusFields, ignoreSericeAccountName, ignoreTimeout, sortTaskRuns); d != "" {
+		t.Fatalf("Did not get expected TaskRuns: %s", diff.PrintWantGot(d))
+	}
+
+	t.Logf("Successfully finished test TestPipelineRunMatrixed")
```

Comment:
This is quite an extensive testing done, thanks a bunch for this detailed test case.

NIT: An additional or alternative way of testing this run can be done with comparing  "status.childReferences[]".